### PR TITLE
[Net] Allow branch-specific MultiplayerAPIs.

### PR DESCRIPTION
**Pull request by Faless**
Removes custom_multiplayer from Node.
MultiplayerAPI overrides are now set at SceneTree level, and apply to whole branches.
Impact on performance when using only the default multiplayer or overriding it is minimal, the use of branches can likely be further optimized by caching nodes and relevant MultiplayerAPI IDs.

~Draft since based on #57646 .~
~Draft because of: https://github.com/godotengine/godot/pull/57647#issuecomment-1109963930~

See demo project: [proj_branch.zip](https://github.com/godotengine/godot/files/8037440/proj_branch.zip)

```gdscript
# main.gd
extends Control

@onready
var server = $combo/server/game
@onready
var client = $combo/client/game

func _ready():
	get_tree().set_multiplayer(MultiplayerAPI.new(), $combo/server.get_path())
	get_tree().set_multiplayer(MultiplayerAPI.new(), $combo/client.get_path())
	server.start_server()
	client.start_client()
```

```gdscript
# game.gd
extends Control

const PORT = 4343
const HOST = "127.0.0.1"

func _bind():
	multiplayer.peer_connected.connect(_connected)


func _connected(id):
	$Panel/VBoxContainer/RichTextLabel.append_text("\nConnected %d" % id)


func start_server():
	_bind()
	var peer = ENetMultiplayerPeer.new()
	peer.create_server(PORT)
	get_multiplayer().multiplayer_peer = peer


func start_client():
	_bind()
	var peer = ENetMultiplayerPeer.new()
	peer.create_client(HOST, PORT)
	get_multiplayer().multiplayer_peer = peer


func _on_button_pressed():
	ping.rpc()


@rpc(any_peer)
func ping():
	$Panel/VBoxContainer/RichTextLabel.append_text("\nPing from %d" % multiplayer.get_remote_sender_id())
```

branch_screen

### CODE REVIEW

File: doc/classes/SceneTree.xml [hunk 1]
```diff
@@ -99,6 +99,13 @@
 				Returns the current frame number, i.e. the total frame count since the application started.
 			</description>
 		</method>
+		<method name="get_multiplayer" qualifiers="const">
+			<return type="MultiplayerAPI" />
+			<argument index="0" name="for_path" type="NodePath" default="NodePath(&quot;&quot;)" />
```
Comment (akien-mga) [inline doc/classes/SceneTree.xml:104]: Should be `root_path` like in the description?

**Review by akien-mga — APPROVED** [review]
Seems good aside from documentation nitpick.

**Review by mhilbrunner — APPROVED** [review]

**Comment by Faless** [comment]
There seem to be a memory leak at exit when using the MultiplayerAPI, **but the issue seem to also affect master**:
```
WARNING: ObjectDB instances leaked at exit (run with --verbose for details).
     at: cleanup (core/object/object.cpp:2031)
Leaked instance: RefCounted:-9222567744099057283
Leaked instance: ENetMultiplayerPeer:-9222568843610685058
Leaked instance: ENetConnection:-9222568293854871167
Leaked instance: SceneReplicationInterface:-9222600729447890549
Leaked instance: MultiplayerAPI:-9222601279203704434
Leaked instance: SceneCacheInterface:-9222599080180448267
Leaked instance: SceneRPCInterface:-9222599629936262126
Leaked instance: RefCounted:-9222600179692075952
```

If fear this might introduce more of that so I've put this in draft again before it gets merged (since it has been approved).
I'll report back and possibly fix the leak in this PR.

**Comment by Faless** [comment]
Nevermind, this seems related to MultiplayerAPI internals and not its usage in SceneTree, will investigate and fix in a dedicated PR.

**Pull request merged**